### ISSUE TITLE
Examples for parsing aggregation results

### ISSUE THREAD

**Issue by vrecan** [olivere/elastic#285]
Please use the following questions as a guideline to help me answer
your issue/question without further inquiry. Thank you.

### Which version of Elastic are you using?
Elastic 1.5.2

### Please describe the expected behavior
Simple examples for parsing types out of elastic results. I'm doing a term's aggregation query on a few fields, some are strings and some are ip's. The same code needs to parse both types and return them into strings. 

### Please describe the actual behavior
NA

### Any steps to reproduce the behavior?
NA

**Comment by vrecan** [comment]
Example function

```
func (l ES) termAggStringByHourLogsIndex(field string, value string, startTime time.Time, aggFields ...string) (map[string][]string, int64, error) {
	keysFromFields := make(map[string][]string, 0)
	index := "logs-" + startTime.Format("2006-01-02")
	endTime := startTime.Add(1 * time.Hour)
	// log.Debug("START: ", startTime.Format("2006-01-02T15:04:05Z"))
	// log.Debug("ENDTIME: ", endTime.Format("2006-01-02T15:04:05Z"))
	// log.Debug("INDEX: ", index)
	rf := E.NewRangeFilter("normalDate").Gte(startTime.Format("2006-01-02T15:04:05Z")).Lt(endTime.Format("2006-01-02T15:04:05Z"))
	tq := E.NewTermQuery(field, value)
	bq1 := E.NewBoolQuery().Must(rf, tq)
	s := l.es.Search().Index(index).Query(bq1)
	for _, aField := range aggFields {
		tagg := E.NewTermsAggregation().Field(aField)
		s = s.Aggregation(aField, tagg)
	}
	s.Size(0)
	r, err := s.Do()
	if nil != err {
		return keysFromFields, 0, err
	}
	// SPEW.Dump(r)

	for _, aField := range aggFields {
		ranges, ok := r.Aggregations.Terms(aField)
		if !ok {
			continue
		}
		for _, res := range ranges.Buckets {
			_, ok := keysFromFields[aField]
			if !ok {
				v := make([]string, 0)
				v = append(v, res.Key.(string))
				keysFromFields[aField] = v
			}
		}
	}
	return keysFromFields, r.TookInMillis, nil
```

**Comment by olivere** [comment]
You might take a look at the test code, search_aggs_test.go in particular.

**Comment by vrecan** [comment]
Yea I did look at it, it still didn't make any sense to me

**Comment by olivere** [comment]
It might help to review the raw response from ES.

If ES returns strings in the aggs (res.Key in your code), you can cast to a string. If it returns a number, you can cast to float64. You should be the one who knows what ES will return; Elastic simply deserializes the JSON response. If you must accept anything, you probably need a type switch.

**Comment by vrecan** [comment]
yea I started doing that but in es responses you get a helper field called key_as_string but that doesn't seem to exist on our side. So It's just key float64 which is actually an ip address so I have to turn a float into an ip address which is annoying

**Comment by vrecan** [comment]
so I see that some things have a helper method with KeyAsString but not the terms aggregation. It's there in the json so if I create a struct to serialize into it should work? 

BTW thanks for your help on this!

**Comment by olivere** [comment]
Can you post the query and the raw output from ES please? I'll have a look at it tomorrow.

There already is KeyAsString for some aggregation types, but not for Terms aggs.

**Comment by vrecan** [comment]
How do I get the raw json that will be constructed?

**Comment by vrecan** [comment]
I can't see how to take a fully built search request and turn it into json

**Comment by vrecan** [comment]
if I follow the examples I should just be able to marshal them but It doesn't work with the output from this:
```
s := Search().Index(index).Query(bq1)
src := s.Source(true)

log.Info(string(json.Marshal(src)))
```
The output is always an empty json `{}`

**Comment by olivere** [comment]
@vrecan Easiest is to [enable the trace log](https://github.com/olivere/elastic/wiki/Logging) and paste its output.

**Comment by olivere** [comment]
I just released 2.0.44 and 3.0.38 which add deserialization of the `key_as_string` field in the buckets of a Terms aggregation. You might then use `res.KeyAsString` to access the stringified key as returned by ES (as e.g. in the buckets of the Histogram aggregation).

**Comment by vrecan** [comment]
@olivere Awesome, thanks!
